content/courses/advanced-ada/parts/resource_management/anonymous_access_types.rst

```diff
@@ -760,15 +760,321 @@ using anonymous access types in the declaration of the :ada:`B` and :ada:`A`
+
+Note that we're not using the :ada:`in` parameter mode in the
+:ada:`Show` procedure above. Usually, this parameter mode can be omitted,
+as it is the default parameter mode |mdash| :ada:`procedure P (I : Integer)`
+is the same as :ada:`procedure P (I : in Integer)`. However, in the case of
+the :ada:`Show` procedure, the :ada:`in` parameter mode isn't just optionally
+absent. In fact, for access parameters, the parameter mode is always implied
+as :ada:`in`, so writing it explicitly is actually forbidden. In other words,
+we can only write :ada:`N : access String` or
+:ada:`N : access constant String`, but we cannot write
+:ada:`N : in access String` or :ada:`N : in access constant String`.
+
+.. admonition:: For further reading...
+
+    When we discussed
+    :ref:`parameters as access values <Adv_Ada_Parameters_As_Access_Values>`
+    in the previous chapter, we saw how we can simply use different
+    parameter modes to write a program instead of using access types.
+    Basically, to implement the same functionality, we just replaced the access
+    types by selecting the correct parameter modes instead and used *simpler*
+    data types.
+
+    Let's do the same exercise again, this time by adapting the previous code
+    example with anonymous access types:
+
+    .. code:: ada run_button project=Courses.Advanced_Ada.Resource_Management.Anonymous_Access_Types.Anonymous_Access_Parameters.Names_String
+
+        package Names is
+
+           function Init (S1, S2 : String)
+                          return String;
+
+           procedure Show (N : String);
+
+        private
+
+           function Init (S1, S2 : String)
+                          return String is
+             (S1 & "-" & S2);
+
+        end Names;
+
+        with Ada.Text_IO; use Ada.Text_IO;
+
+        package body Names is
+
+           procedure Show (N : String) is
+           begin
+              Put_Line ("Name: " & N);
+           end Show;
+
+        end Names;
+
+        with Names; use Names;
+
+        procedure Show_Names is
+           N : String := Init ("Lily", "Ann");
+        begin
+           Show (N);
+        end Show_Names;
+
+    Although we're using simple strings instead of access types in this version
+    of the code example, we're still getting a similar behavior. However, there
+    is a small, yet important difference in the way the string returned by
+    :ada:`Init` is being allocated: while the previous implementation (which
+    was using an access result type) was allocating the string on the heap,
+    we're now allocating the string on the stack.
+
+Later on, we talk about the
+:ref:`accessibility rules in the case of access parameters <Adv_Ada_Accessibility_Rules_Access_Parameters>`.
+
+In general, we should avoid access parameters whenever possible and simply use
+objects and parameter modes directly, as it makes the design simpler and less
+error-prone. The only exception is when we're interfacing to other languages,
```

Comment:
The other time access parameters are useful -- vital -- is for inherited primitive operations. The inherited operation will have the new child type designated but will still dynamically dispatch. Otherwise, if we used a named access type designating the parent type, that formal parameter would still have the same named access type in the inherited operation, rather than one designating the child type.

Reply:
Thanks, this is indeed a very good point! I've just integrated your comment into a new subsection about this topic.

One question, though: is this issue affecting only tagged types, or could we talk about derived types in general? In my code examples, I see that *untagged* types have a similar behavior...

Reply:
Yes, that's true, it isn't only an issue for tagged types. I guess I think of it more for tagged types because of dynamic dispatching, which is only possible for primitive operations, but you're right that it is a general consideration.